singer/src/main/java/com/pinterest/singer/writer/s3/S3Writer.java

```diff
@@ -0,0 +1,438 @@
+import com.pinterest.singer.thrift.LogMessageAndPosition;
+import com.pinterest.singer.thrift.configuration.S3WriterConfig;
+import com.pinterest.singer.utils.SingerUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.PosixFilePermission;
+import java.nio.file.attribute.PosixFilePermissions;
+
+import java.util.Set;
+import java.util.List;
+import java.util.Date;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.text.SimpleDateFormat;
+
+import software.amazon.awssdk.services.s3.S3Client;
+
+import static com.pinterest.singer.utils.SingerUtils.getHostnamePrefixes;
+
+/**
+ * A LogStreamWriter for Singer that writes to S3 (writer.type=s3).
+ * */
+public class S3Writer implements LogStreamWriter {
+    public static final String HOSTNAME = SingerUtils.HOSTNAME;
+    private static final Logger LOG = LoggerFactory.getLogger(S3Writer.class);
+    private static final SimpleDateFormat formatter = new SimpleDateFormat("yyyyMMddHHmmssSSS");
+    protected final LogStream logStream;
+    private final String BUFFER_DIR;
+    private final int mB = 1024 * 1024;
+    private final int MIN_UPLOAD_TIME_IN_SECONDS = 30;
+    private final int MAX_FILE_SIZE_IN_MB = 50;
+
+    private ObjectUploaderTask putObjectUploader;
+    private S3Client s3Client;
+    private final S3WriterConfig s3WriterConfig;
+
+    // S3 information
+    private String bucketName;
+    private String keyPrefix;
+
+    // Disk-buffered file that will eventually be uploaded to S3 if size or time thresholds are met
+    private BufferedOutputStream bufferedOutputStream;
+    private File bufferFile;
+
+    // Custom Thresholds
+    private int maxFileSizeMB;
+    private int minUploadTime;
+    private int maxRetries;
+
+    // Timer for scheduling uploads
+    private static ScheduledExecutorService fileUploadTimer;
+    private Future<?> uploadFuture;
+    private final Object objLock = new Object(); // used for synchronization locking
+
+    private int messageCount = 0;
+
+    static {
+        ScheduledThreadPoolExecutor tmpTimer = new ScheduledThreadPoolExecutor(1);
+        tmpTimer.setRemoveOnCancelPolicy(true);
+        fileUploadTimer = tmpTimer;
+    }
+
+    /**
+     * Constructs an S3Writer instance.
+     *
+     * @param logStream the LogStream associated with this writer
+     * @param s3WriterConfig the S3WriterConfig containing configuration settings
+     */
+    public S3Writer(LogStream logStream, S3WriterConfig s3WriterConfig) {
+        this.logStream = logStream;
+        this.s3WriterConfig = s3WriterConfig;
+        this.BUFFER_DIR = "/tmp/singer/s3";
```

Comment:
Should this be customizable via configs? Just in case this path is not available or cannot be written to for some reason.

Reply:
Yes, I have just added a config. Thanks Vahid!